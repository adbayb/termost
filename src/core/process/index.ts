import { spawn } from "child_process";

export const exec = async (command: string, options: ExecOptions = {}) => {
	const promise = new Promise<string>((resolve, reject) => {
		let stdout = "";
		let stderr = "";
		const [bin, ...args] = command.split(" ") as [string, ...string[]];

		const childProcess = spawn(bin, args, {
			cwd: options.cwd,
			shell: true,
			stdio: "pipe",
			env: {
				...process.env,
				// @note: make sure to force color display for spawned processes
				FORCE_COLOR: "1",
			},
		});

		childProcess.stdout.on("data", (chunk) => {
			stdout += chunk;
		});

		childProcess.stderr.on("data", (chunk) => {
			stderr += chunk;
		});

		childProcess.on("close", (exitCode) => {
			if (exitCode !== 0) {
				reject(stderr.trim());
			} else {
				resolve(stdout.trim());
			}
		});
	});

	try {
		return await promise;
	} catch (error) {
		throw new Error(error);
	}
};

type ExecOptions = {
	cwd?: string;
};
