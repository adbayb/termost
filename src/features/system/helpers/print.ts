import chalk from "chalk";

type FormatColor =
	| "red"
	| "green"
	| "yellow"
	| "blue"
	| "magenta"
	| "cyan"
	| "grey"
	| "black"
	| "white";

type FormatModifier =
	| "bold"
	| "italic"
	| "underline"
	| "strikethrough"
	| "uppercase"
	| "lowercase";

const chalkByFormatColor = {
	red: "redBright",
	green: "greenBright",
	yellow: "yellowBright",
	blue: "blueBright",
	magenta: "magentaBright",
	cyan: "cyanBright",
	grey: "blackBright",
	black: "black",
	white: "white",
} as const;

const chalkByModifier = {
	bold: "bold",
	italic: "italic",
	underline: "underline",
	strikethrough: "strikethrough",
} as const;

type FormatFunction<ReturnValue = string> = (
	message: string,
	options?: {
		color?: FormatColor;
		modifier?: FormatModifier | Array<FormatModifier>;
	}
) => ReturnValue;

export const format: FormatFunction = (message, { color, modifier } = {}) => {
	let transformer: chalk.Chalk = chalk;

	if (color) {
		transformer = transformer[chalkByFormatColor[color]];
	}

	if (modifier) {
		const processOneModifier = (mod: FormatModifier) => {
			if (mod === "uppercase") {
				message = message.toUpperCase();
			} else if (mod === "lowercase") {
				message = message.toLowerCase();
			} else {
				transformer = transformer[chalkByModifier[mod]];
			}
		};

		if (Array.isArray(modifier)) {
			for (const mod of modifier) {
				processOneModifier(mod);
			}
		} else {
			processOneModifier(modifier);
		}
	}

	return transformer(message);
};

export const print: FormatFunction<void> = (message, options) => {
	console.log(format(message, options));
};

type BannerType = "success" | "error" | "warning" | "information";

const bannerPropertiesByType: Record<
	BannerType,
	{ icon: string; color: FormatColor }
> = {
	error: {
		icon: "❌",
		color: "red",
	},
	information: {
		icon: "ℹ️ ",
		color: "blue",
	},
	success: {
		icon: "✅",
		color: "green",
	},
	warning: {
		icon: "⚠️ ",
		color: "yellow",
	},
};

export const banner = (type: BannerType, message: string) => {
	const { icon, color } = bannerPropertiesByType[type];

	print(`${icon} ${message}`, { color, modifier: "bold" });
};
