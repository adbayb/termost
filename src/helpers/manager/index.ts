import { CommandName } from "../../features/types";
import { createQueue } from "./queue";

export const getManager = (commandName: CommandName): Manager => {
	const manager = managerCollection[commandName];

	if (!manager) {
		throw new Error(
			`No manager has been instantiated for the \`${commandName}\` command`
		);
	}

	return manager;
};

export const createManager = (commandName: CommandName): Manager => {
	const instructions = createQueue<Instruction>();

	const manager: Manager = {
		// addValue,
		// addOption,
		// getValue,
		// getOption,
		addInstruction(instruction) {
			instructions.enqueue(instruction);
		},
		async run() {
			while (!instructions.isEmpty()) {
				const task = instructions.dequeue();

				if (task) {
					await task();
				}
			}
		},
	};

	managerCollection[commandName] = manager;

	return manager;
};

type Instruction = () => Promise<void>;

type Manager = {
	addInstruction(instruction: Instruction): void;
	run(): Promise<void>;
};

const managerCollection: Record<CommandName, Manager> = {};
