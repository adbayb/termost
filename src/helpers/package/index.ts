import fs from "fs";
import path from "path";
import { PackageMetadata } from "../../types";

export const findNearestPackageJson = (pathname: string): PackageMetadata => {
	const packagePathname = path.resolve(pathname, "package.json");

	if (isFileExists(packagePathname)) {
		return require(packagePathname);
	}

	return findNearestPackageJson(path.resolve(pathname, ".."));
};

const isFileExists = (pathname: string) => {
	try {
		fs.accessSync(pathname, fs.constants.F_OK);

		return true;
	} catch {
		return false;
	}
};

export const getPackageMetadata = () => {
	const mainFilename = require.main?.filename;

	if (!mainFilename) {
		throw new Error(
			"Termost was unable to retrieve automatically the package name and version. To fix it, use `termost({ name, description, version })` to define them manually."
		);
	}

	return findNearestPackageJson(path.dirname(mainFilename));
};
