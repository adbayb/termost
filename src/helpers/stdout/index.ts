import chalk from "chalk";

/**
 * A helper to format an arbitrary text as a print input
 * @param message The text to display
 * @param options The configuration object to control the formatting properties
 * @returns The formatted text
 */
export const format = (
	message: string,
	{
		color,
		modifier,
	}: {
		color?: Color;
		modifier?: Modifier | Array<Modifier>;
	} = {}
): string => {
	let transformer: chalk.Chalk = chalk;

	if (color) {
		transformer = transformer[chalkByFormatColor[color]];
	}

	if (modifier) {
		const processOneModifier = (mod: Modifier) => {
			if (mod === "uppercase") {
				message = message.toUpperCase();
			} else if (mod === "lowercase") {
				message = message.toLowerCase();
			} else {
				transformer = transformer[chalkByModifier[mod]];
			}
		};

		if (Array.isArray(modifier)) {
			for (const mod of modifier) {
				processOneModifier(mod);
			}
		} else {
			processOneModifier(modifier);
		}
	}

	return transformer(message);
};

/**
 * An opinionated helper to display arbitrary text on the console
 * @param message The text to display. Use an array if you need to display a message in several lines
 * @param options The configuration object to define the display type and/or override the default label
 */
export const print = (
	message: string | Array<string>,
	{ label, type = "information" }: { type?: MessageType; label?: string } = {}
) => {
	const { color, defaultLabel, icon, method } = formatPropertiesByType[type];
	const messages = typeof message === "string" ? [message] : message;

	method(
		format(`\n${icon} ${label || defaultLabel}`, {
			color,
			modifier: "bold",
		})
	);

	for (const message of messages) {
		method(format(`   ${message}`, { color }));
	}
};

const formatPropertiesByType = {
	error: {
		color: "red",
		defaultLabel: "Error",
		icon: "❌",
		method: console.error,
	},
	information: {
		color: "blue",
		defaultLabel: "Information",
		icon: "ℹ️ ",
		method: console.info,
	},
	success: {
		color: "green",
		defaultLabel: "Success",
		icon: "✅",
		method: console.log,
	},
	warning: {
		color: "yellow",
		defaultLabel: "Warning",
		icon: "⚠️ ",
		method: console.warn,
	},
} as const;

const chalkByFormatColor = {
	red: "redBright",
	green: "greenBright",
	yellow: "yellowBright",
	blue: "blueBright",
	magenta: "magentaBright",
	cyan: "cyanBright",
	grey: "blackBright",
	black: "black",
	white: "white",
} as const;

const chalkByModifier = {
	bold: "bold",
	italic: "italic",
	underline: "underline",
	strikethrough: "strikethrough",
} as const;

type MessageType = "success" | "error" | "warning" | "information";

type Color =
	| "red"
	| "green"
	| "yellow"
	| "blue"
	| "magenta"
	| "cyan"
	| "grey"
	| "black"
	| "white";

type Modifier =
	| "bold"
	| "italic"
	| "underline"
	| "strikethrough"
	| "uppercase"
	| "lowercase";
