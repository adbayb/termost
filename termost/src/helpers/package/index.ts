import process from "node:process";
import { resolve } from "node:path";
import { createRequire } from "node:module";
import { accessSync, constants } from "node:fs";

import type { PackageMetadata } from "../../types";

const binCallerPath = process.argv[1] ?? "";
const require = createRequire(binCallerPath);

export const getPackageMetadata = (
	pathname: string = binCallerPath,
): PackageMetadata => {
	try {
		const packagePathname = resolve(pathname, "package.json");

		if (isFileExists(packagePathname)) {
			return require(packagePathname) as PackageMetadata;
		}

		return getPackageMetadata(resolve(pathname, ".."));
	} catch (error) {
		throw new Error(
			`Termost was unable to retrieve automatically the package name and version. To fix it, use \`termost({ name, description, version })\` to define them manually.\nMore details: ${String(error)}.`,
		);
	}
};

const isFileExists = (pathname: string) => {
	try {
		accessSync(pathname, constants.F_OK);

		return true;
	} catch {
		return false;
	}
};
