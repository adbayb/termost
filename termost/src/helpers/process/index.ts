import process from "node:process";
import { spawn } from "node:child_process";

export const exec = async (command: string, options: ExecOptions = {}) => {
	const { cwd, hasLiveOutput = false } = options;

	return new Promise<string>((resolve, reject) => {
		let stdout = "";
		let stderr = "";
		const [bin, ...arguments_] = command.split(" ") as [string, ...string[]];

		// eslint-disable-next-line sonarjs/os-command
		const childProcess = spawn(bin, arguments_, {
			cwd,
			env: {
				// eslint-disable-next-line n/no-process-env
				...process.env,
				// @note: make sure to force color display for spawned processes
				FORCE_COLOR: "1",
			},
			shell: true,
			stdio: hasLiveOutput ? "inherit" : "pipe",
		});

		childProcess.stdout?.on("data", (chunk: string) => {
			stdout += chunk;
		});

		childProcess.stderr?.on("data", (chunk: string) => {
			stderr += chunk;
		});

		childProcess.on("close", (exitCode) => {
			if (exitCode !== 0) {
				const output = `${stderr}${stdout}`;

				reject(new Error(output.trim()));
			} else {
				resolve(stdout.trim());
			}
		});
	});
};

type ExecOptions = {
	cwd?: string;
	hasLiveOutput?: boolean;
};
