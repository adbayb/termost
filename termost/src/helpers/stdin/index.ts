import process from "node:process";

export const getArguments = () => {
	const parameters = process.argv.slice(2);
	let command: string | undefined;
	const operands: string[] = [];
	const options: Record<string, boolean | number | string> = {};
	let currentOptionName: string | undefined;

	const addOptimisticOption = (name: string, value?: boolean | string) => {
		if (value) {
			options[name] = typeof value === "string" ? castValue(value) : true;
		} else {
			currentOptionName = name;
		}
	};

	const flushOptimisticOption = () => {
		if (currentOptionName) {
			options[currentOptionName] = true;
			currentOptionName = undefined;
		}
	};

	for (const parameter of parameters) {
		const shortFlagMatchResult = SHORT_FLAG_REGEX.exec(parameter)?.groups;
		const longFlagMatchResult = LONG_FLAG_REGEX.exec(parameter)?.groups;

		if (shortFlagMatchResult?.name) {
			flushOptimisticOption();

			const optionFlags = [...shortFlagMatchResult.name];
			const lastIndex = optionFlags.length - 1;

			optionFlags.forEach((flag, index) => {
				addOptimisticOption(
					flag,
					lastIndex === index ? undefined : true,
				);
			});
		} else if (longFlagMatchResult?.name) {
			flushOptimisticOption();
			addOptimisticOption(
				longFlagMatchResult.name,
				longFlagMatchResult.value,
			);
		} else if (currentOptionName) {
			options[currentOptionName] = castValue(parameter);
			currentOptionName = undefined;
		} else if (!command) {
			command = parameter;
		} else {
			operands.push(parameter);
		}
	}

	flushOptimisticOption();

	return { command, operands, options };
};

const SHORT_FLAG_REGEX = /^-(?<name>(?!-).*)$/;
const LONG_FLAG_REGEX = /^--(?<name>.*?)(?:=(?<value>.+))?$/;

const castValue = (value: string) => {
	try {
		return JSON.parse(value) as boolean | number | string;
	} catch {
		return value;
	}
};
