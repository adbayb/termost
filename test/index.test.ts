import { exec } from "../src/helpers/process";

jest.setTimeout(20000);

describe("termost", () => {
	test("should display `help`", async () => {
		const longFlagOutput = await exec(
			"node -r esbuild-register example/index.ts --help"
		);
		const shortFlagOutput = await exec(
			"node -r esbuild-register example/index.ts -h"
		);

		expect(longFlagOutput).toMatchSnapshot();
		expect(shortFlagOutput).toMatchSnapshot();
	});

	test("should handle `command` api", async () => {
		const helpOutput = await exec(
			"node -r esbuild-register example/withCommand.ts --help"
		);
		const buildOutput = await exec(
			"node -r esbuild-register example/withCommand.ts build"
		);
		const watchOutput = await exec(
			"node -r esbuild-register example/withCommand.ts watch"
		);
		const buildSharedFlagOutput = await exec(
			"node -r esbuild-register example/withCommand.ts build --global --local hello"
		);
		const watchSharedFlagOutput = await exec(
			"node -r esbuild-register example/withCommand.ts watch --global"
		);
		const buildHelpOutput = await exec(
			"node -r esbuild-register example/withCommand.ts build --help"
		);
		const watchHelpOutput = await exec(
			"node -r esbuild-register example/withCommand.ts watch --help"
		);

		expect(helpOutput).toMatchSnapshot();
		expect(buildOutput).toMatchSnapshot();
		expect(watchOutput).toMatchSnapshot();
		expect(buildSharedFlagOutput).toMatchSnapshot();
		expect(watchSharedFlagOutput).toMatchSnapshot();
		expect(buildHelpOutput).toMatchSnapshot();
		expect(watchHelpOutput).toMatchSnapshot();
	});

	test("should handle `option` api", async () => {
		const output = await exec(
			"node -r esbuild-register example/withOption.ts"
		);

		expect(output).toMatchSnapshot();
	});

	test("should handle `task` api", async () => {
		const output = await exec(
			"node -r esbuild-register example/withTask.ts"
		);

		expect(output).toMatchSnapshot();
	});
});
