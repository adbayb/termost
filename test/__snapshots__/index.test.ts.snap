// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`termost should display \`help\` 1`] = `
"[93m[1m[4m[24m[22m[39m
[93m[1m[4mUSAGE:[24m[22m[39m
[92mtermost[39m [...options]
[93m[1m[4m[24m[22m[39m
[93m[1m[4mDESCRIPTION:[24m[22m[39m
Program description placeholder. Program name and version are retrieved from your \`package.json\`. You can override this automatic retrieval by using the \`termost({ name, description, version })\` builder form.
[93m[1m[4m[24m[22m[39m
[93m[1m[4mOPTIONS:[24m[22m[39m
  [92m-h, --help    [39m Display the help center
  [92m-v, --version [39m Print the version
  [92m-f, --flag    [39m A super useful CLI flag"
`;

exports[`termost should display \`help\` 2`] = `
"[93m[1m[4m[24m[22m[39m
[93m[1m[4mUSAGE:[24m[22m[39m
[92mtermost[39m [...options]
[93m[1m[4m[24m[22m[39m
[93m[1m[4mDESCRIPTION:[24m[22m[39m
Program description placeholder. Program name and version are retrieved from your \`package.json\`. You can override this automatic retrieval by using the \`termost({ name, description, version })\` builder form.
[93m[1m[4m[24m[22m[39m
[93m[1m[4mOPTIONS:[24m[22m[39m
  [92m-h, --help    [39m Display the help center
  [92m-v, --version [39m Print the version
  [92m-f, --flag    [39m A super useful CLI flag"
`;

exports[`termost should handle \`command\` api 1`] = `
"[93m[1m[4m[24m[22m[39m
[93m[1m[4mUSAGE:[24m[22m[39m
[92mtermost[39m <command> [...options]
[93m[1m[4m[24m[22m[39m
[93m[1m[4mDESCRIPTION:[24m[22m[39m
Example to showcase the \`command\` API
[93m[1m[4m[24m[22m[39m
[93m[1m[4mCOMMANDS:[24m[22m[39m
  [92mbuild         [39m Transpile and bundle in production mode
  [92mwatch         [39m Rebuild your assets on any code change
[93m[1m[4m[24m[22m[39m
[93m[1m[4mOPTIONS:[24m[22m[39m
  [92m-h, --help    [39m Display the help center
  [92m-v, --version [39m Print the version
  [92m--global      [39m Shared flag between commands"
`;

exports[`termost should handle \`command\` api 2`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëã Hello, I'm the build command[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Shared global flag = false[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Local command flag = local-value[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Context value = {\\"command\\":\\"build\\",\\"values\\":{\\"globalFlag\\":false,\\"localFlag\\":\\"local-value\\"}}[39m"
`;

exports[`termost should handle \`command\` api 3`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëã Hello, I'm the watch command[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Shared global flag = false[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Context value = {\\"command\\":\\"watch\\",\\"values\\":{\\"globalFlag\\":false}}[39m"
`;

exports[`termost should handle \`command\` api 4`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëã Hello, I'm the build command[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Shared global flag = true[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Local command flag = hello[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Context value = {\\"command\\":\\"build\\",\\"values\\":{\\"globalFlag\\":true,\\"localFlag\\":\\"hello\\"}}[39m"
`;

exports[`termost should handle \`command\` api 5`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëã Hello, I'm the watch command[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Shared global flag = true[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   üëâ Context value = {\\"command\\":\\"watch\\",\\"values\\":{\\"globalFlag\\":true}}[39m"
`;

exports[`termost should handle \`command\` api 6`] = `
"[93m[1m[4m[24m[22m[39m
[93m[1m[4mUSAGE:[24m[22m[39m
[92mtermost build[39m [...options]
[93m[1m[4m[24m[22m[39m
[93m[1m[4mDESCRIPTION:[24m[22m[39m
Transpile and bundle in production mode
[93m[1m[4m[24m[22m[39m
[93m[1m[4mOPTIONS:[24m[22m[39m
  [92m-h, --help    [39m Display the help center
  [92m-v, --version [39m Print the version
  [92m--global      [39m Shared flag between commands
  [92m--local       [39m Local command flag"
`;

exports[`termost should handle \`command\` api 7`] = `
"[93m[1m[4m[24m[22m[39m
[93m[1m[4mUSAGE:[24m[22m[39m
[92mtermost watch[39m [...options]
[93m[1m[4m[24m[22m[39m
[93m[1m[4mDESCRIPTION:[24m[22m[39m
Rebuild your assets on any code change
[93m[1m[4m[24m[22m[39m
[93m[1m[4mOPTIONS:[24m[22m[39m
  [92m-h, --help    [39m Display the help center
  [92m-v, --version [39m Print the version
  [92m--global      [39m Shared flag between commands"
`;

exports[`termost should handle \`message\` api 1`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   A content formatted thanks to the \`print\` helper presets.[39m
[92m[1m[22m[39m
[92m[1m‚úÖ Success[22m[39m
[92m   A content formatted thanks to the \`print\` helper presets.[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  üëã You can also customize the label[22m[39m
[94m   A content formatted thanks to the \`print\` helper presets.[39m
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  üëã You can also customize the label[22m[39m
[94m   I support also[39m
[94m   multilines[39m
[94m   with array input[39m
[37m[3m[9m[1m[22m[29m[23m[39m
[37m[3m[9m[1mYou can also have a total control on the formatting through the \`format\` helper.[22m[29m[23m[39m"
`;

exports[`termost should handle \`option\` api 1`] = `
"[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   {[39m
[94m  \\"optionWithAlias\\": 0,[39m
[94m  \\"optionWithoutAlias\\": \\"defaultValue\\"[39m
[94m}[39m"
`;

exports[`termost should handle \`task\` api 1`] = `
"[STARTED] Task with returned value (persisted)
[SUCCESS] Task with returned value (persisted)
[STARTED] Task with side-effect only (no persisted value)
[SUCCESS] Task with side-effect only (no persisted value)
[STARTED] Task can also access other persisted task values
[SUCCESS] Task can also access other persisted task values
[STARTED] Or even execute external commands thanks to its provided helpers
[SUCCESS] Or even execute external commands thanks to its provided helpers
[STARTED] A task can have a dynamic label generated from contextual values: small
[SUCCESS] A task can have a dynamic label generated from contextual values: small
[94m[1m[22m[39m
[94m[1m‚ÑπÔ∏è  Information[22m[39m
[94m   A task with a specified \\"key\\" can be retrieved here. Size = 45. If no \\"key\\" was specified the task returned value cannot be persisted across program instructions.[39m
{
  \\"size\\": 45,
  \\"computedFromOtherTaskValues\\": \\"small\\",
  \\"execOutput\\": \\"Hello from shell\\"
}"
`;
